Add support for incrementing version number

na-ka-na opened the pull request:
Adding support for a 'version' number column which increments on every update. Useful for a few reasons:

* Easier to see how many updates have been made to a row
* Adding primary keys to the history table. E.g. if the main table has a primary key 'id', it will allow adding a primary key 'id', 'version' to the history table. A lot of ORMs expect a primary key

na-ka-na commented:
> Hello again, I tried out your method changes and everything is looking good code-wise. Although it is less than ideal to have so many procedural calls to the database, we are planning to mitigate this in a different way.
> 
> In the meantime, would you mind updating your examples in the README to reflect the existing example table and structure? This should make it easier for folks to try it out.
> 
> Thanks for your contribution!

Thank you I'll address all the comments soon and update the PR

na-ka-na commented on versioning_function.sql:
```diff
@@ -210,16 +250,31 @@ BEGIN
-          array_to_string(commonColumns, ',$1.') ||
-          ',tstzrange($2, $3, ''[)''))'
-        USING OLD, range_lower, time_stamp_to_use;
+        IF increment_version = 'true' THEN
```
Sounds good

na-ka-na commented on README.md:
````diff
@@ -369,6 +369,48 @@ When migration mode is enabled:
+To achieve this:
+* Add an `int` `version` column (or any other name you prefer) to the base table, e.g.
+   ```sql
+   ALTER TABLE your_table ADD COLUMN version int NOT NULL DEFAULT 1
````
Done

na-ka-na commented on README.md:
````diff
@@ -369,6 +369,48 @@ When migration mode is enabled:
+   ```
+* Add the same to the history table
+   ```sql
+   ALTER TABLE your_table_history ADD COLUMN version int NOT NULL
````
Done

na-ka-na commented on README.md:
```diff
@@ -369,6 +369,48 @@ When migration mode is enabled:
+   CREATE TRIGGER versioning_trigger
+   BEFORE INSERT OR UPDATE OR DELETE ON your_table
+   FOR EACH ROW EXECUTE PROCEDURE versioning(
+     'sys_period', 'your_table_history', false, false, false, false,
```
Done

na-ka-na commented on README.md:
````diff
@@ -369,6 +369,48 @@ When migration mode is enabled:
+
+After this, if you insert a new row
+```sql
+INSERT INTO your_table (id, value) VALUES ('my_id', 'abc')
````
Done

na-ka-na commented on README.md:
````diff
@@ -369,6 +369,48 @@ When migration mode is enabled:
+```sql
+INSERT INTO your_table (id, value) VALUES ('my_id', 'abc')
+```
+the table will start with the row having the initial version `id=my_id, value=abc, version=1`.
````
Done

na-ka-na commented on README.md:
````diff
@@ -369,6 +369,48 @@ When migration mode is enabled:
+
+If then, the row gets updated with
+```sql
+UPDATE your_table SET value='def' WHERE id='my_id'
````
Done

na-ka-na commented on README.md:
````diff
@@ -369,6 +369,48 @@ When migration mode is enabled:
+```sql
+UPDATE your_table SET value='def' WHERE id='my_id'
+```
+then the table will reflect incremented version `id=my_id, value=def, version=2`. And correspondingly the history table will have the old version `id=my_id, value=abc, version=1` (or both versions if `include_current_version_in_history` is turned on).
````
Done

na-ka-na commented on versioning_function.sql:
```diff
@@ -59,6 +62,8 @@ BEGIN
   ignore_unchanged_values := COALESCE(TG_ARGV[3],'false');
   include_current_version_in_history := COALESCE(TG_ARGV[4],'false');
   enable_migration_mode := COALESCE(TG_ARGV[5],'false');
+  increment_version := COALESCE(TG_ARGV[6],'false');
```
Not done since I'm using the established pattern. But this is a good suggestion

na-ka-na commented on versioning_function.sql:
```diff
@@ -59,6 +62,8 @@ BEGIN
   ignore_unchanged_values := COALESCE(TG_ARGV[3],'false');
   include_current_version_in_history := COALESCE(TG_ARGV[4],'false');
   enable_migration_mode := COALESCE(TG_ARGV[5],'false');
+  increment_version := COALESCE(TG_ARGV[6],'false');
```
Not done since I'm using the established pattern. But this is a good suggestion

na-ka-na commented on versioning_function.sql:
```diff
@@ -88,6 +93,19 @@ BEGIN
   END IF;
 
+  -- check version column
+  IF increment_version = 'true' THEN
```
Not done since I'm using the established pattern. But this is a good suggestion

na-ka-na commented on versioning_function.sql:
```diff
@@ -210,16 +250,31 @@ BEGIN
-          array_to_string(commonColumns, ',$1.') ||
-          ',tstzrange($2, $3, ''[)''))'
-        USING OLD, range_lower, time_stamp_to_use;
+        IF increment_version = 'true' THEN
```
Discussed here https://github.com/nearform/temporal_tables/pull/94/files#r2140986643

mscottx88 commented:
Thank you so much for applying those changes to documentation. After further review, I notice that there are no updates to the existing tests. To continue to pull this change into the repository, we will need to have tests updated and added. Let me know if you need any guidance on how the tests function okay? Thanks!

na-ka-na commented:
> Thank you so much for applying those changes to documentation. After further review, I notice that there are no updates to the existing tests. To continue to pull this change into the repository, we will need to have tests updated and added. Let me know if you need any guidance on how the tests function okay? Thanks!

Will do!

mscottx88 commented:
Thank you so much! I'll take a look tonight!

On Tue, Jul 1, 2025, 7:06 AM Simone Busoli ***@***.***> wrote:

> @simoneb <https://github.com/simoneb> requested your review on: #94
> <https://github.com/nearform/temporal_tables/pull/94> Add support for
> incrementing version number.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nearform/temporal_tables/pull/94#event-18412142477>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE7IYMJR3M7EV3KJ72XXIRL3GKIVNAVCNFSM6AAAAAB7DRIBS2VHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJYGQYTEMJUGI2DONY>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

mscottx88 reviewed (approved):
Approved thank you for your contributions!

simoneb reviewed (changes requested):
Thanks for the contribution, appreciate all the effort that went into it.

I'd like to ask for a couple of changes to the readme:

1. can you please add, at the top of the readme, where we mention the list of newly added features, this new feature with an internal hyperlink to the section where you describe the feature?
2. the place where you added the documentation for the feature has been added _below_ the anchor tag for the migrations section (which in all honesty, I'm not even sure is needed). Anyway, please move it _above_ the migrations anchor

na-ka-na commented:
> Thanks for the contribution, appreciate all the effort that went into it.
> 
> I'd like to ask for a couple of changes to the readme:
> 
> 1. can you please add, at the top of the readme, where we mention the list of newly added features, this new feature with an internal hyperlink to the section where you describe the feature?
> 2. the place where you added the documentation for the feature has been added _below_ the anchor tag for the migrations section (which in all honesty, I'm not even sure is needed). Anyway, please move it _above_ the migrations anchor

Done. I moved it over the ### Migrating to include_current_version_in_history section. Lemme know if you meant otherwise

simoneb commented:
> Done. I moved it over the ### Migrating to include_current_version_in_history section. Lemme know if you meant otherwise

Almost. Considering that the last new feature we added was "include current version in history", I would consider that section of the docs and the section "migrating to include current version in history" to go together, therefore I would suggest moving the section of the docs that you added _below_ "migrating to include current version in history".

I hope it makes sense.

na-ka-na commented:
> > Done. I moved it over the ### Migrating to include_current_version_in_history section. Lemme know if you meant otherwise
> 
> Almost. Considering that the last new feature we added was "include current version in history", I would consider that section of the docs and the section "migrating to include current version in history" to go together, therefore I would suggest moving the section of the docs that you added _below_ "migrating to include current version in history".
> 
> I hope it makes sense.

Ok done. I think that's where I had it earlier, so I changed it back
